# Recommender Systems

https://medium.com/@sumanadhikari/building-a-movie-recommendation-engine-using-scikit-learn-8dbb11c5aa4b

In [4]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [5]:
df = pd.read_csv("movie_dataset.csv")

In [3]:
df[1:5] # examining the data

index     budget                            genres  \
1      1  300000000          Adventure Fantasy Action   
2      2  245000000            Action Adventure Crime   
3      3  250000000       Action Crime Drama Thriller   
4      4  260000000  Action Adventure Science Fiction   

                                       homepage      id  \
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
1  ocean drug abuse exotic island east india trad...                en   
2         spy based on novel secret agent sequel mi6                en   
3  dc comics crime fighter terrorist secret ident...                en   
4  based on novel mars medallion space travel pri...                en   

                             original_title  \
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  ... runtime  \
1  Captain Barbossa, long believed to be dead, ha...  139.082615  ...   169.0   
2  A cryptic message from Bond’s past sends him o...  107.376788  ...   148.0   
3  Following the death of District Attorney Harve...  112.312950  ...   165.0   
4  John Carter is a war-weary, former military ca...   43.926995  ...   132.0   

                                    spoken_languages    status  \
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count  \
1  Pirates of the Caribbean: At World's End          6.9       4500   
2                                   Spectre          6.3       4466   
3                     The Dark Knight Rises          7.6       9106   
4                               John Carter          6.1       2124   

                                                cast  \
1  Johnny Depp Orlando Bloom Keira Knightley Stel...   
2  Daniel Craig Christoph Waltz L\u00e9a Seydoux ...   
3  Christian Bale Michael Caine Gary Oldman Anne ...   
4  Taylor Kitsch Lynn Collins Samantha Morton Wil...   

                                                crew           director  
1  [{'name': 'Dariusz Wolski', 'gender': 2, 'depa...     Gore Verbinski  
2  [{'name': 'Thomas Newman', 'gender': 2, 'depar...         Sam Mendes  
3  [{'name': 'Hans Zimmer', 'gender': 2, 'departm...  Christopher Nolan  
4  [{'name': 'Andrew Stanton', 'gender': 2, 'depa...     Andrew Stanton  

[4 rows x 24 columns]

In [6]:
features = ['keywords', 'cast', 'genres', 'director'] # we dont need all of the features

In [8]:
def combine_features(row):
    return row['keywords'] + ' ' + row['cast'] + ' ' + row["genres"] + ' ' + row['director']

Object `row` not found.


In [10]:
for feature in features:
    df[feature] = df[feature].fillna('') # to fill all the black NaNs with ea blank string
df["combined_features"] = df.apply(combine_features, axis=1)

In [12]:
cv = CountVectorizer()
count_matrix = cv.fit_transform(df["combined_features"])
cosine_sim = cosine_similarity(count_matrix)
cosine_sim[0,:]

array([1.        , 0.10540926, 0.12038585, ..., 0.        , 0.        ,
       0.        ])

In [15]:
cosine_sim.shape

(4803, 4803)

(consine_siume[0,:] a list of how simailar 0 is to every other 4803 movies in the list. 1 is the same its the id is 0 so its the first element)

In [16]:
cosine_sim.shape

(4803, 4803)

In [18]:
def get_title_from_index(index):
    return df[df.index == index]['title'].values[0]

In [21]:
def get_index_from_title(title):
    return df[df.title == title]['index'].values[0]

In [28]:
movie_user_likes = "Men in Black 3" # hardcoded input here
movie_index = get_index_from_title(movie_user_likes)
similar_movies = list(enumerate(cosine_sim[movie_index]))

In [29]:
sorted_similar_movies = sorted(similar_movies, key=lambda x: x[1], reverse=True)[1:]

In [30]:
i = 0
print("Top 5 similar movies to " + movie_user_likes + " are: \n")
for element in sorted_similar_movies:
    print(get_title_from_index(element[0]))
    i=i+1
    if i>=5:
        break

Top 5 similar movies to Men in Black 3 are: 

Men in Black II
Men in Black
The Helix... Loaded
Time Changer
Hot Tub Time Machine


In [36]:
df['genres']

0        Action Adventure Fantasy Science Fiction
1                        Adventure Fantasy Action
2                          Action Adventure Crime
3                     Action Crime Drama Thriller
4                Action Adventure Science Fiction
5                        Fantasy Action Adventure
6                                Animation Family
7                Action Adventure Science Fiction
8                        Adventure Fantasy Family
9                        Action Adventure Fantasy
10       Adventure Fantasy Action Science Fiction
11                Adventure Action Thriller Crime
12                       Adventure Fantasy Action
13                       Action Adventure Western
14       Action Adventure Fantasy Science Fiction
15                       Adventure Family Fantasy
16               Science Fiction Action Adventure
17                       Adventure Action Fantasy
18                  Action Comedy Science Fiction
19                       Action Adventure Fantasy


In [27]:
df.columns

Index(['index', 'budget', 'genres', 'homepage', 'id', 'keywords',
       'original_language', 'original_title', 'overview', 'popularity',
       'production_companies', 'production_countries', 'release_date',
       'revenue', 'runtime', 'spoken_languages', 'status', 'tagline', 'title',
       'vote_average', 'vote_count', 'cast', 'crew', 'director',
       'combined_features'],
      dtype='object')